ruby/ql/test/query-tests/security/cwe-327/BrokenCryptoAlgorithm.expected

```diff
@@ -0,0 +1,18 @@
+| broken_crypto.rb:4:8:4:34 | call to new | The cryptographic algorithm DES is broken or weak, and should not be used. |
+| broken_crypto.rb:8:1:8:18 | call to update | The cryptographic algorithm DES is broken or weak, and should not be used. |
+| broken_crypto.rb:12:8:12:43 | call to new | The cryptographic algorithm AES is broken or weak, and should not be used. |
```

Comment:
It's a bit of a shame that this message doesn't mention the specific cipher mode that makes it weak. Would that be difficult to add?

Reply:
Yeah, the inaccurate message here is not great. At the level of `CryptographicOperation`, we have no concept of the block mode, so the weakness/strength of the operation is just a kind of black box. The solutions that I've considered are:
 1. Keeping `CryptographicAlgorithm` and the block mode separate, but introducing a proper `BlockMode` class with e.g. an `isWeak()` predicate of its own. `CryptographicOperation` could then have a `BlockMode getBlockMode()` predicate rather than its own `isWeak()` predicate, and the query could check both the encryption algorithm and the block mode.
 2. Include the block mode in the `CryptographicAlgorithm`, and add 2 predicates - `isWeakEncryptionAlgorithm` and `isWeakBlockMode`, to that class. Then `CryptographicOperation#isWeak()` could be removed.

I think I prefer `1` overall as it seems like the intention of the `CryptographicAlgorithm` class (shared with JS and Python) is to represent the algorithm modulo any block mode. Do you have any preferences between these?

Reply:
1 sounds like a nice solution, if it's not too difficult to implement.